### PR TITLE
Disable the X-XSS-Protection header in defaults

### LINKED ISSUE #35: [Security Issue] `X-XSS-Protection` header
Based on this [OWASP article](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#x-xss-protection-header), it is recommended to disable the `X-XSS-Protection` header by setting `X-XSS-Protection: 0` to avoid security issues in older browsers.

* Currently, the default is `:xss-protection {:enable? true, :mode :block}` at [this line](https://github.com/ring-clojure/ring-defaults/blob/4ecb31d97be116b3c177c651c013cd2834bb528a/src/ring/middleware/defaults.clj#L47).

* Also, setting `:xss-protection {:enable? false}` raises an error on the pre-condition check for `nil` at [this line](https://github.com/ring-clojure/ring-headers/blob/master/src/ring/middleware/x_headers.clj#L94), because the code sends an empty map [here](https://github.com/ring-clojure/ring-defaults/blob/master/src/ring/middleware/defaults.clj#L81) (`(dissoc options :enable?)` => `{}`).

I suggest updating [x-headers in the ring-headers repository](https://github.com/ring-clojure/ring-headers/blob/master/src/ring/middleware/x_headers.clj) to enforce this solution on all the dependent repositories. However, if you would prefer only fixing this repo, I can create a PR for it.

Thanks for your attention.

### PR DESCRIPTION
**Pull request by terop**
Disable the XSS Auditor in older browsers by default. The
X-XSS-Protection header has been deprecated by modern browsers due to
security issues it introduces on the client.

See: https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#x-xss-protection-header

Fixes #35.

### CODE REVIEW

**Comment by terop** [comment]
I resurrected the change from PR #36 with the requested commit message change. Because I cannot updated the existing the PR I created a new one instead.

**Comment by weavejester** [comment]
Thanks for the update. However, now that I take another look at the code of this PR, there are parts that confuse me. If the X-XSS-Protection header is deprecated, why is it still set to be added to the response? In addition, the link in the commit is now broken.

Let's instead just remove the option from site-defaults, and add a deprecation notice to the README. Let's reference this URL for the reason for the change in the commit message: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-XSS-Protection

**Comment by terop** [comment]
I tried to do the changes you suggested but I am not sure if they are correct and sufficient.

File: src/ring/middleware/defaults.clj [hunk 1]
```diff
@@ -78,7 +77,9 @@
         args))
 
 (defn- wrap-xss-protection [handler options]
-  (x/wrap-xss-protection handler (:enable? options true) (dissoc options :enable?)))
+  (x/wrap-xss-protection handler
+                         (:enable? options false)
+                         (-> options (dissoc :enable?) not-empty)))
```
Comment (weavejester) [inline src/ring/middleware/defaults.clj:82]: Are these changes still necessary? What happens if we take them out?

**Comment by weavejester** [comment]
The changes look good, and were exactly what I was thinking of. I'm not sure what the original PR's change to lines 80-82 was about. What happens if we revert those alterations?

File: src/ring/middleware/defaults.clj (hunk 1 above)
Comment (terop) [reply src/ring/middleware/defaults.clj]: Hard to say, at least unit tests still pass.

**Comment by terop** [comment]
Code changed as suggested, unit tests pass at least.

**Comment by terop** [comment]
@weavejester could you kindly check the latest changes?

File: README.md [hunk 2]
```diff
@@ -138,8 +138,9 @@ The following configuration keys are supported:
     server is on a non-standard port. See: [wrap-ssl-redirect][9].
 
   - `:xss-protection` -
-    Enable the X-XSS-Protection header that tells supporting browsers
-    to use heuristics to detect XSS attacks. See: [wrap-xss-protection][10].
+    **Deprecated.** Enable the X-XSS-Protection header. This is no
+    longer considered best practice by OWASP and should be avoided.
```
Comment (weavejester) [inline README.md:142]: Let's add a link around `no longer considered best practice` that links to the MDN page on it, and lets remove `by OWASP` since the OWASP page makes no mention of it now.

File: test/ring/middleware/defaults_test.clj [hunk 3]
```diff
@@ -99,13 +97,11 @@
       (is (= (set (keys (:headers resp)))
              #{"X-Frame-Options"
                "X-Content-Type-Options"
-               "X-XSS-Protection"
                "Strict-Transport-Security"
                "Content-Type"
                "Set-Cookie"}))
       (is (= (get-in resp [:headers "X-Frame-Options"]) "SAMEORIGIN"))
       (is (= (get-in resp [:headers "X-Content-Type-Options"]) "nosniff"))
-      (is (= (get-in resp [:headers "X-XSS-Protection"]) "1; mode=block"))
```
Comment (weavejester) [inline test/ring/middleware/defaults_test.clj:108]: Can you add a small unit test to check that if X-XSS-Protection is on, it actually still works? If we haven't removed the key, we should check to ensure it hasn't been broken.

**Review by weavejester — CHANGES_REQUESTED** [review]
Good work! I have some comments around adding a link to the README to justify the change, and a unit test to ensure that it's not broken, only deprecated.

File: README.md (hunk 2 above)
Comment (terop) [reply README.md]: Changed as suggested.

File: test/ring/middleware/defaults_test.clj (hunk 3 above)
Comment (terop) [reply test/ring/middleware/defaults_test.clj:108]: Unit test added.

**Comment by terop** [comment]
@weavejester could you again kindly take a look at the latest changes?

File: README.md [hunk 4]
```diff
@@ -138,8 +138,10 @@ The following configuration keys are supported:
     server is on a non-standard port. See: [wrap-ssl-redirect][9].
 
   - `:xss-protection` -
-    Enable the X-XSS-Protection header that tells supporting browsers
-    to use heuristics to detect XSS attacks. See: [wrap-xss-protection][10].
+    **Deprecated** Enable the X-XSS-Protection header. This is [no
+    longer considered best practice](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-XSS-Protection)
```
Comment (weavejester) [inline README.md:142]: Can you use the "footnote" syntax for this link, to make it consistent with the rest of the README? So:

    This is [no longer considered best practice][13]

Comment (terop) [reply README.md]: Changed as suggested.

**Comment by weavejester** [comment]
Perfect! Thank you.

**Comment by terop** [comment]
@weavejester would it be possible to soon have a new release of ring-defaults with this change?

**Comment by weavejester** [comment]
Done. Apologies for the delay.

**Comment by terop** [comment]
No problem, many thanks for the release.
